kolibri/utils/version.py

```diff
@@ -125,21 +103,139 @@ def parse_git_tag_version_string(version_string):
     )
+    timestamp = git_log.communicate()[0]
+    try:
+        timestamp = datetime.datetime.utcfromtimestamp(int(timestamp))
+    except ValueError:
+        return None
+    return timestamp.strftime('%Y%m%d%H%M%S')
+
+
+def get_version_file():
+    """
+    Looks for a file VERSION in the package data and returns the contents in
+    this. Does not check consistency.
+    """
+    return pkgutil.get_data('kolibri', 'VERSION').decode('utf-8')
+
+
+def get_prerelease_version(version):
+    """
+    Called when kolibri.VERSION is set to a non-final version:
+
+    if version ==
+    *, *, *, "alpha", 0: Maps to latest commit timestamp
+    *, *, *, "alpha", >0: Uses latest git tag, asserting that there is such.
+    """
+
+    major = get_major_version(version)
+
+    mapping = {'alpha': 'a', 'beta': 'b', 'rc': 'c'}
+    sub = ''
+
+    tag_describe = git_describe()
+    if tag_describe:
+        git_version = get_version_from_git(tag_describe)
+        if version[4] == 0:
+            # If it's a alpha/beta==0, it might not even be tagged, so we
+            # don't call assert_git_version
+            if version[3] == 'alpha':
+                mapping['alpha'] = '.dev'
+                sub = get_git_changeset()
+            else:
+                sub = '.dev{}{}'.format(*git_version[4:])
+        else:
+            # When the version is > 0, it has to be tagged
+            assert_git_version(version, git_version)
+            if git_version[0] > 0:
+                # If there's commits beyond the tag, automatically append
+                # a .dev
+                sub = '.dev{}{}'.format(*git_version[4:])
+        if git_version[5]:
+            sub += '-' + git_version[5]
+    else:
+        # No git data, will look for a VERSION file
+        version_file = get_version_file()
+        if version_file:
+            # TODO: This only checks major digits, should also check
+            # release string (alpha/beta/rc/dev etc)
+            if not version_file.startswith(major):
+                raise AssertionError(
+                    "VERSION file inconsistent with kolibri.__version__. "
+                    "__version__ is: {}, File says: {}".format(
+                        str(version),
+                        version_file
+                    )
+                )
+            else:
+                return version_file
+
+    sub = mapping[version[3]] + str(version[4]) + sub
+
+    return str(major + sub)
+
+
+@lru_cache()
+def get_version(version=None):
+    """
+    Returns a PEP 440-compliant version number from VERSION.
+    Derives additional information from git repository if code is contained
+    in such.
+
+    This is important to read from PEP-404 (which this function is compliant
+    with):
+
+    Within a numeric release ( 1.0 , 2.7.3 ), the following suffixes are
+    permitted and MUST be ordered as shown:
+
+    .devN, aN, bN, rcN, <no suffix>, .postN
+    """
+    version = get_complete_version(version)
```

Comment:
This seems to be going further than fixing the issue of the `VERSION` file sometimes be missing, and seems to be treating the set `VERSION` in `kolibri.__init__.py` as canonical, rather than as a fallback in case a version file is missing and git tags are unavailable.